PermissionError: [WinError 32] The process cannot access the file because it is being used by another process in AppVeyor

* **I'm submitting a ...**

- [x] bug report
- [ ] feature request
- [ ] question about the decisions made in the repository

* **Do you want to request a *feature* or report a *bug*?**

bug


* **What is the current behavior?**

Session tests seem to try removing session files before the server, which holds them open is shut down. This raises the exception on Windows in random CI jobs:
```python-traceback
================================== FAILURES ===================================
_________________________ SessionTest.test_3_Redirect _________________________

self = <cherrypy.test.test_session.SessionTest testMethod=test_3_Redirect>

    def tearDown(self):
        # Clean up sessions.
        for fname in os.listdir(localDir):
            if fname.startswith(sessions.FileSession.SESSION_PREFIX):
                path = Path(localDir) / fname
>               path.remove_p()

cherrypy\test\test_session.py:146: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.tox\python\lib\site-packages\path.py:1323: in remove_p
    self.unlink()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = Path('C:\\projects\\cherrypy\\cherrypy\\test\\session-e9e9bd7b2315a0155f961d6b6f1546489d4c09e0.lock')

    def unlink(self):
        """ .. seealso:: :func:`os.unlink` """
>       os.unlink(self)
E       PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: Path('C:\\projects\\cherrypy\\cherrypy\\test\\session-e9e9bd7b2315a0155f961d6b6f1546489d4c09e0.lock')

.tox\python\lib\site-packages\path.py:1332: PermissionError
------------------------------ Captured log call ------------------------------
_cplogging.py              284 INFO     127.0.0.1 - - [04/Sep/2018:19:55:05] "GET /testStr HTTP/1.1" 200 1 "" ""
_cplogging.py              284 INFO     127.0.0.1 - - [04/Sep/2018:19:55:05] "GET /redir_target HTTP/1.1" 200 11 "" ""
______________________ SessionTest.test_4_File_deletion _______________________

self = <cherrypy.test.test_session.SessionTest testMethod=test_4_File_deletion>

    def tearDown(self):
        # Clean up sessions.
        for fname in os.listdir(localDir):
            if fname.startswith(sessions.FileSession.SESSION_PREFIX):
                path = Path(localDir) / fname
>               path.remove_p()

cherrypy\test\test_session.py:146: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.tox\python\lib\site-packages\path.py:1323: in remove_p
    self.unlink()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = Path('C:\\projects\\cherrypy\\cherrypy\\test\\session-e9e9bd7b2315a0155f961d6b6f1546489d4c09e0.lock')

    def unlink(self):
        """ .. seealso:: :func:`os.unlink` """
>       os.unlink(self)
E       PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: Path('C:\\projects\\cherrypy\\cherrypy\\test\\session-e9e9bd7b2315a0155f961d6b6f1546489d4c09e0.lock')

.tox\python\lib\site-packages\path.py:1332: PermissionError
------------------------------ Captured log call ------------------------------
_cplogging.py              284 INFO     127.0.0.1 - - [04/Sep/2018:19:55:06] "GET /testStr HTTP/1.1" 200 1 "" ""
_cplogging.py              284 INFO     127.0.0.1 - - [04/Sep/2018:19:55:06] "GET /testStr HTTP/1.1" 200 1 "" ""
_______________________ SessionTest.test_5_Error_paths ________________________

self = <cherrypy.test.test_session.SessionTest testMethod=test_5_Error_paths>

    def tearDown(self):
        # Clean up sessions.
        for fname in os.listdir(localDir):
            if fname.startswith(sessions.FileSession.SESSION_PREFIX):
                path = Path(localDir) / fname
>               path.remove_p()

cherrypy\test\test_session.py:146: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.tox\python\lib\site-packages\path.py:1323: in remove_p
    self.unlink()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = Path('C:\\projects\\cherrypy\\cherrypy\\test\\session-e9e9bd7b2315a0155f961d6b6f1546489d4c09e0.lock')

    def unlink(self):
        """ .. seealso:: :func:`os.unlink` """
>       os.unlink(self)
E       PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: Path('C:\\projects\\cherrypy\\cherrypy\\test\\session-e9e9bd7b2315a0155f961d6b6f1546489d4c09e0.lock')

.tox\python\lib\site-packages\path.py:1332: PermissionError

```

* **If the current behavior is a bug, please provide the steps to reproduce and if possible a screenshots and logs of the problem. If you can, show us your code.**

https://ci.appveyor.com/project/CherryPy/cherrypy/branch/master/job/5jnboghrrlc0mvyj#L776


* **What is the expected behavior?**

All file descriptors released before removal.

* **What is the motivation / use case for changing the behavior?**

Tests should be stable and deterministic.

* **Please tell us about your environment:**

- Cheroot version: N/A
- CherryPy version: master
- Python version: 3.6.X
- OS: Windows (AppVeyor)
- Browser: N/A

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, e.g. stackoverflow, gitter, etc.)

https://github.com/pallets/werkzeug/issues/1264#issuecomment-391461840 (related, maybe our client also forgets to close connection?)
https://stackoverflow.com/a/2023709/595220 (a way of tracking what exactly holds the file descriptor open)

Similar traceback @ https://ci.appveyor.com/project/CherryPy/cherrypy/build/1.0.410/job/g32c0qhi9h9t9a5i/tests:
```python-traceback
self = <cherrypy.test.test_static.StaticTest testMethod=test_unicode>

    @pytest.mark.xfail(py27_on_windows, reason='#1544')  # noqa: E301
    def test_unicode(self):
        ensure_unicode_filesystem()
        with self.unicode_file():
            url = ntou('/static/Слава Україні.html', 'utf-8')
            # quote function requires str
            url = tonative(url, 'utf-8')
            url = urllib.parse.quote(url)
            self.getPage(url)
    
            expected = ntou('Героям Слава!', 'utf-8')
>           self.assertInBody(expected)

cherrypy\test\test_static.py:427: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
c:\program files\python\Lib\contextlib.py:66: in __exit__
    next(self.gen)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    @staticmethod
    @contextlib.contextmanager
    def unicode_file():
        filename = ntou('Слава Україні.html', 'utf-8')
        filepath = os.path.join(curdir, 'static', filename)
        with io.open(filepath, 'w', encoding='utf-8') as strm:
            strm.write(ntou('Героям Слава!', 'utf-8'))
        try:
            yield
        finally:
>           os.remove(filepath)
E           PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\projects\\cherrypy\\cherrypy\\test\\static\\Слава Україні.html'

cherrypy\test\test_static.py:410: PermissionError
```

--- a/cherrypy/test/test_session.py
+++ b/cherrypy/test/test_session.py
@@ -139,8 +139,10 @@ def setup_server():
 class SessionTest(helper.CPWebCase):
     setup_server = staticmethod(setup_server)
 
-    def tearDown(self):
-        # Clean up sessions.
+    @classmethod
+    def teardown_class(cls):
+        """Clean up sessions."""
+        super(cls, cls).teardown_class()
         consume(
             file.remove_p()
             for file in localDir.listdir()

--- a/cherrypy/test/test_static.py
+++ b/cherrypy/test/test_static.py
@@ -152,8 +152,9 @@ class StaticTest(helper.CPWebCase):
         vhost = cherrypy._cpwsgi.VirtualHost(rootApp, {'virt.net': testApp})
         cherrypy.tree.graft(vhost)
 
-    @staticmethod
-    def teardown_server():
+    @classmethod
+    def teardown_class(cls):
+        super(cls, cls).teardown_class()
         for f in (has_space_filepath, bigfile_filepath):
             f.remove_p()
 

--- a/cherrypy/test/test_static.py
+++ b/cherrypy/test/test_static.py
@@ -53,6 +53,7 @@ BIGFILE_SIZE = 32 * MB
 
 
 class StaticTest(helper.CPWebCase):
+    files_to_remove = []
 
     @staticmethod
     def setup_server():
@@ -155,7 +156,9 @@ class StaticTest(helper.CPWebCase):
     @classmethod
     def teardown_class(cls):
         super(cls, cls).teardown_class()
-        for f in (has_space_filepath, bigfile_filepath):
+        files_to_remove = has_space_filepath, bigfile_filepath
+        files_to_remove += tuple(cls.files_to_remove)
+        for f in files_to_remove:
             f.remove_p()
 
     def test_static(self):
@@ -395,17 +398,13 @@ class StaticTest(helper.CPWebCase):
         self.getPage('/static/\x00')
         self.assertStatus('404 Not Found')
 
-    @staticmethod
-    @contextlib.contextmanager
-    def unicode_file():
+    @classmethod
+    def unicode_file(cls):
         filename = ntou('Слава Україні.html', 'utf-8')
-        filepath = os.path.join(curdir, 'static', filename)
-        with io.open(filepath, 'w', encoding='utf-8') as strm:
+        filepath = curdir / 'static' / filename
+        with filepath.open('w', encoding='utf-8')as strm:
             strm.write(ntou('Героям Слава!', 'utf-8'))
-        try:
-            yield
-        finally:
-            os.remove(filepath)
+        cls.files_to_remove.append(filepath)
 
     py27_on_windows = (
         platform.system() == 'Windows' and
@@ -414,7 +413,8 @@ class StaticTest(helper.CPWebCase):
     @pytest.mark.xfail(py27_on_windows, reason='#1544')  # noqa: E301
     def test_unicode(self):
         ensure_unicode_filesystem()
-        with self.unicode_file():
+        self.unicode_file()
+        if True:
             url = ntou('/static/Слава Україні.html', 'utf-8')
             # quote function requires str
             url = tonative(url, 'utf-8')

--- a/cherrypy/test/test_static.py
+++ b/cherrypy/test/test_static.py
@@ -1,5 +1,4 @@
 # -*- coding: utf-8 -*-
-import contextlib
 import io
 import os
 import sys

--- a/cherrypy/test/test_static.py
+++ b/cherrypy/test/test_static.py
@@ -414,15 +414,14 @@ class StaticTest(helper.CPWebCase):
     def test_unicode(self):
         ensure_unicode_filesystem()
         self.unicode_file()
-        if True:
-            url = ntou('/static/Слава Україні.html', 'utf-8')
-            # quote function requires str
-            url = tonative(url, 'utf-8')
-            url = urllib.parse.quote(url)
-            self.getPage(url)
+        url = ntou('/static/Слава Україні.html', 'utf-8')
+        # quote function requires str
+        url = tonative(url, 'utf-8')
+        url = urllib.parse.quote(url)
+        self.getPage(url)
 
-            expected = ntou('Героям Слава!', 'utf-8')
-            self.assertInBody(expected)
+        expected = ntou('Героям Слава!', 'utf-8')
+        self.assertInBody(expected)
 
 
 def error_page_404(status, message, traceback, version):

--- a/cherrypy/test/test_static.py
+++ b/cherrypy/test/test_static.py
@@ -10,6 +10,7 @@ from http.client import HTTPConnection
 
 import pytest
 import py.path
+import path
 
 import cherrypy
 from cherrypy.lib import static
@@ -40,9 +41,9 @@ def ensure_unicode_filesystem():
         tmpdir.remove()
 
 
-curdir = os.path.join(os.getcwd(), os.path.dirname(__file__))
-has_space_filepath = os.path.join(curdir, 'static', 'has space.html')
-bigfile_filepath = os.path.join(curdir, 'static', 'bigfile.log')
+curdir = path.Path(__file__).dirname()
+has_space_filepath = curdir / 'static' / 'has space.html'
+bigfile_filepath = curdir / 'static' / 'bigfile.log'
 
 # The file size needs to be big enough such that half the size of it
 # won't be socket-buffered (or server-buffered) all in one go. See
@@ -154,11 +155,7 @@ class StaticTest(helper.CPWebCase):
     @staticmethod
     def teardown_server():
         for f in (has_space_filepath, bigfile_filepath):
-            if os.path.exists(f):
-                try:
-                    os.unlink(f)
-                except Exception:
-                    pass
+            f.remove_p()
 
     def test_static(self):
         self.getPage('/static/index.html')

